rdflib/plugins/shared/jsonld/util.py

```diff
@@ -17,22 +17,31 @@
 
 from urllib.parse import urljoin, urlsplit, urlunsplit
 
-from rdflib.parser import create_input_source, PythonInputSource
+from rdflib.parser import create_input_source, PythonInputSource, StringInputSource
 
-from io import StringIO
+from io import TextIOBase, TextIOWrapper
 
 
 def source_to_json(source):
 
     if isinstance(source, PythonInputSource):
         return source.data
 
+    if isinstance(source, StringInputSource):
+        return json.load(source.getCharacterStream())
+
     # TODO: conneg for JSON (fix support in rdflib's URLInputSource!)
     source = create_input_source(source, format="json-ld")
 
     stream = source.getByteStream()
     try:
-        return json.load(StringIO(stream.read().decode("utf-8")))
+        # Use character stream as-is, or interpret byte stream as UTF-8
+        if isinstance(stream, TextIOBase):
```

Comment:
I'm actually unsure if this condition will be true given this code from `rdflib.parser.FileInputSource`:

https://github.com/RDFLib/rdflib/blob/9379a69d6ec6e18819aa0c5a0d54849e7abb8223/rdflib/parser.py#L239-L244

Reply:
FWIW, I was trying to home in on the error that was triggered by trying to parse the output from `json.dump`, which specifically caused the stream result from `source.getByteStream()` to be a character stream.  I agree that seems not quite right, and the problem may lie deeper in the logic that sorts out the data stream to be parsed.  

But I was also trying to, identify a minimal fix that could solve my immediate problem.  Hopefully if someone chooses to delve deeper into the `create_input_source` logic, the test cases provided will still be helpful.